ZNE in Catalyst blog post (#604)
unitaryfund/unitary.fund: src/content/blog/2024_zne_catalyst.md

---

> title: "Just in time for just-in-time error mitigation"

Reviewer: Minor comment, but as the title is quite important for SEO, could be good to give it a 'subtitle'. E.g.,

> Just in time for just-in-time error mitigation: compiling error mitigation with LLVM, X, and Y.

Could sprinkle in important keywords as well like Mitiq, Catalyst, etc.

author: Alessandro Cosentino
day: 31
month: 10
year: 2024

> ---

Reviewer 2: Please add tags. See the blog post for a list, please include Mitiq.

We all want faster computer programs. Just-in-time (JIT) compilation has been a key concept in classical compiler design to help achieve faster computer programs. Most modern programming languages and runtime environments support some kind of JIT optimization techniques. In fact, very recently, an [experimental JIT compiler](https://docs.python.org/3/whatsnew/3.13.html#whatsnew313-jit-compiler) was added to Python in its latest release, 3.13.

> 

Reviewer: minor, but it could be worth (very briefly!) explaining what a JIT is, as readers might not be familiar (could be as simple as saying, at least in this context 'compiling parts of your workflow for performance improvements at execution time, avoiding sometimes slow Python interpretation'.

> Just as JIT compilation accelerates classical programs, similar techniques can be applied to quantum computing to boost performance. We all want faster quantum computer programs.  Catalyst is a JIT compiler framework developed by Xanadu to speed up quantum programs written with their PennyLane library. If you have a PennyLane program, you can decorate it with some simple Catalyst syntax and, in many cases, you have a faster quantum computer program.

Reviewer: ```suggestion
Just as JIT compilation accelerates classical programs, similar techniques can be applied to quantum computing to boost performance. We all want faster quantum computer programs.  [Catalyst](https://github.com/PennyLaneAI/catalyst) is a JIT compiler framework developed by Xanadu to speed up quantum programs written with their PennyLane library. If you have a PennyLane program, you can decorate it with some simple Catalyst syntax and, in many cases, you have a faster quantum computer program.
```

What we also want is more accurate quantum computer programs – accurate enough to be useful for the application or the experiment we seek to run. Quantum error correction is making significant strides, but it still introduces substantial overhead that can be impractical for near-term applications. We look at the more practical techniques of quantum error mitigation to make our quantum programs more accurate. Unitary Fund's own Mitiq project provides a Python package for applying error mitigation techniques to quantum programs. 

> Even though Mitiq can be used with PennyLane programs, it doesn't seamless integrate into the Catalyst framework for JIT compilation. In the last few months, Unitary Fund and Xanadu have partnered to improve the integration of error mitigation techniques into the Catalyst compiler. The natural starting point was Zero-Noise Extrapolation (ZNE), which is

Reviewer 2: ```suggestion
Even though [Mitiq](https://github.com/unitaryfund/mitiq) could already be used with PennyLane programs, it didn't seamlessly integrate into the Catalyst framework for JIT compilation. In the last few months, Unitary Fund and Xanadu have partnered to improve the integration of error mitigation techniques, upstreaming them directly from Mitiq into the Catalyst compiler. The natural starting point was Zero-Noise Extrapolation (ZNE), which is
```

Reviewer 2: this is the first time Mitiq is cited, in the blog post, I think it should be introduced in a previous line.

Author: It's not true that now Mitiq itself integrates into Catalyst, that's why I didn't use the past tense. It's also not true that we upstreamed techniques directly from Mitiq. What we did is rewriting those techniques from scratch in the Catalyst framework.
This has been a matter of confusion since the initial proposal, and it's something I have highlighted in the final report as well.

> the most popular error mitigation technique in Mitiq, and arguably the simplest to implement, yet very effective.

Reviewer 2: ```suggestion
the most popular error mitigation technique in Mitiq, and arguably the simplest to implement, yet it can be very effective.
```

- **Stage 1**: Executing the program at increased noise levels
- **Stage 2**: Extrapolating back to the zero-noise regime.

If you want to know more about the theory behind ZNE, check out [Mitiq documentation](https://mitiq.readthedocs.io/en/stable/guide/zne-5-theory.html).

We worked on the full Catalyst stack and added the following options to the respective stages of ZNE:

- _Local folding_ – Instead of applying noise scaling globally to the entire circuit, noise level increments can now be applied to individual gates, providing finer control over the noise scaling process.
- _Exponential extrapolation_ – An exponential function is fitted to the execution results of the noise-scaled programs, whereas the previous option only allowed for polynomial extrapolation. Exponential extrapolation can be advantageous in certain noise models where the error compounds exponentially with noise scaling.

Why is this important? It turns out that, depending on the underlying noise models, 
some folding techniques and some fitting techniques may perform better than others. 

> In complex noise environments, having the flexibility to choose the most effective folding and extrapolation methods can significantly improve error mitigation outcomes. The more complex the noise model is, the more impactful it is to have such a level of customization in ZNE. Should this blog post inspire you to become a better ZNE practitioner, check out [this review paper](https://arxiv.org/abs/2307.05203) for how to make the most out of all the ZNE options.

Reviewer 2: Can use the citations tool for notes.

Author: done

After implementing these options, it was time to test them. Catalyst programs can already be run on Amazon Braket and error mitigation can be leveraged to attenuate the hardware noise there. Running on real hardware can be time- and money-expensive though. 

> We needed a workflow with better feedback loop time, and for that, we turned into adding a simple depolarizing noise model to Unitary Fund's in-house simulator Qrack. If you missed Dan Strano's [post](https://unitary.fund/posts/2024_qrack_catalyst/) on Qrack integration with Catalyst, go and check it out!

Reviewer: What do you mean by feedback loop time here?

Author: What I mean by 'feedback loop time' is the 'time to smoke test the change end-to-end locally'. I call it feedback loop, because the outcome of the smoke test could be something to fix in the code. I will rephrase it.

![Catalyst ZNE diagram.](/images/catalyst-zne-diagram.png)

Did I tell you how much we learned? For the Unitary Fund team, this project was our first venture into the JIT world and the architectural stack that comes with implementing it. 

> We had to ramp up on things like JAX and the MLIR framework (Multi-Level Intermediate Representation). Understanding MLIR gave us deeper insights into compiler design for quantum programs in general as well. I am not going to lie; it was all a bit daunting in the beginning, but it was incredibly rewarding!

Reviewer: Maybe something I am wondering while reading this paragraph, _why_ MLIR/JAX? E.g., if I am not familiar with these technologies, does it bring anything to the table (or is it just another way of doing the same thing as other technologies as QASM?)

---
title: "Just in time for just-in-time error mitigation: Mitiq meets Catalyst"
author: Alessandro Cosentino
day: 18
month: 11
year: 2024
tags:
- Mitiq
- Partnership
---

This blog post is a summary of a partnership between Unitary Fund and 
[Xanadu](https://www.xanadu.ai/) to integrate quantum error mitigation techniques

> from Unitary Fund's Mitiq into Xanadu's PennyLane Catalyst compiler.

Reviewer: @cosenal I wonder if it is worth linking Mitiq and Catalyst websites here, on their first mention? But this is optional, as they are linked right below 🤔

Author: I thought about it, but I prefer below where they are introduced. This paragraph is more like a tl;dr and I don't want to drive readers out of it immediately by clicking on too many links.

## Just-in-time compilation

We all want faster computer programs. Just-in-time (JIT) compilation has been a key concept in classical compiler design to help achieve faster computer programs. The basic idea behind a JIT compiler is that it observes the program running and optimizes it based on those observations. Most modern programming languages and runtime environments nowadays support some kind of JIT optimization techniques. In fact, very recently, an [experimental JIT compiler](https://docs.python.org/3/whatsnew/3.13.html#whatsnew313-jit-compiler) was added to Python in its latest 3.13 release.

> We all want faster quantum computer programs. Just as JIT compilation accelerates classical programs, similar techniques can be applied to quantum computing for boosting performance.   [Catalyst](https://github.com/PennyLaneAI/catalyst) is a JIT compiler framework developed by Xanadu to speed up quantum programs written with their PennyLane library. If you have a PennyLane program, you can easily decorate it with some simple Catalyst syntax and, in many cases, you have a faster quantum computer program.

Reviewer: @cosenal something that is missing from the blog post (but is still interesting/novel) is the fact that the new ZNE transformations _work with structure_ -- that is, they aren't unrolling loops; could be worth adding here, because this means that you can apply error mitigation to very compact program representations.

Author: great suggestion – added in [596b1a6](https://github.com/unitaryfund/unitary.fund/pull/604/commits/596b1a67b23a41aec19f549fc2d7e99da7440f9d)

Why is this important? It turns out that, depending on the underlying noise models, 
some folding techniques and some fitting techniques may perform better than others. 
In complex noise environments, having the flexibility to choose the most effective folding and extrapolation methods can significantly improve error mitigation outcomes. The more complex the noise model is, the more impactful it is to have such a level of customization in ZNE. Should this blog post inspire you to become a better ZNE practitioner, check out [this review paper](https://arxiv.org/abs/2307.05203) for how to make the most out of all the ZNE options.

## Noisy simulation on Catalyst
After implementing these options, it was time to test them. Catalyst programs can already be run on Amazon Braket, and error mitigation can be leveraged to attenuate the hardware noise there. Running on real hardware can be time- and money-expensive though. 

> We needed a workflow to quickly smoke-test the new features, and for that, we turned into adding a simple depolarizing noise model to Unitary Fund's in-house simulator Qrack. If you missed Dan Strano's [post](https://unitary.fund/posts/2024_qrack_catalyst/) on Qrack integration with Catalyst, go and check it out! Qrack is now the first quantum simulator implementing a noise model, as well as integrating with a JIT compiler framework.

Reviewer: nice 😎
